src/neutron/universaladdr.cpp

```diff
@@ -0,0 +1,293 @@
+        }
+        default:
+            // unknown or unsupported version
+            return -5;
+    }
+
+    setVersion(Version(version));
+    setData(data.data() + 2, dataSize);
+    return 0;
+}
+
+int UniversalAddress::setHex(const std::string &str) {
+    return setHex(str.c_str());
+}
+
+std::string UniversalAddress::getHex() {
+    if (m_data.empty()) {
+        return "";
+    }
+    std::stringstream hex;
+    hex << std::hex << std::setfill('0');
+
+    uint16_t version = htole16(m_version);
+    hex << std::setw(2) << static_cast<unsigned>(version & 0xff);
+    hex << std::setw(2) << static_cast<unsigned>(version >> 8);
+
+    for (unsigned char i: m_data) {
+        hex << std::setw(2) << static_cast<unsigned>(i);
+    }
+
+    return hex.str();
+}
+
+class DestinationConverter : public boost::static_visitor<UniversalAddress>
+{
+
+public:
+    UniversalAddress operator()(const CKeyID& id) const {
+        std::vector<unsigned char> data;
+        data.insert(data.end(), id.begin(), id.end());
+        UniversalAddress addr(UniversalAddress::P2PKH, data);
+        return addr;
+    }
+
+    UniversalAddress operator()(const CScriptID& id) const {
+        std::vector<unsigned char> data;
+        data.insert(data.end(), id.begin(), id.end());
+        UniversalAddress addr(UniversalAddress::P2SH, data);
+        return addr;
+    }
+
+    UniversalAddress operator()(const WitnessV0KeyHash& id) const {
+        std::vector<unsigned char> data;
+        data.insert(data.end(), id.begin(), id.end());
+        UniversalAddress addr(UniversalAddress::P2WPKH, data);
+        return addr;
+    }
+
+    UniversalAddress operator()(const WitnessV0ScriptHash& id) const {
+        std::vector<unsigned char> data;
+        data.insert(data.end(), id.begin(), id.end());
+        UniversalAddress addr(UniversalAddress::P2WSH, data);
+        return addr;
+    }
+
+    UniversalAddress operator()(const WitnessUnknown& id) const {
+        return UniversalAddress();
+    }
+
+    UniversalAddress operator()(const X86VMID& id) const {
+        std::vector<unsigned char> data;
+        data.insert(data.end(), id.begin(), id.end());
+        UniversalAddress addr(UniversalAddress::NX86, data);
+        return addr;
+    }
+
+    UniversalAddress operator()(const TestVMID& id) const {
+        std::vector<unsigned char> data;
+        data.insert(data.end(), id.begin(), id.end());
+        UniversalAddress addr(UniversalAddress::NTVM, data);
+        return addr;
+    }
+
+    UniversalAddress operator()(const CNoDestination& no) const {
+        return UniversalAddress();
+    }
+};
+
+UniversalAddress DestinationToUniversal(const CTxDestination& dest) {
+    return boost::apply_visitor(DestinationConverter(), dest);
+}
+
+CTxDestination UniversalToDestination(const UniversalAddress& ua) {
+    switch (ua.version()) {
+        case UniversalAddress::P2PKH:
+        {
+            uint160 hash;
+            if (hash.size() == ua.data().size()) {
```

Comment:
there is a lot of duplicated code here. Is it possible to clean this up some?

Reply:
updated, do u think it is ok ?